### PR TITLE
Make persistent sessions co-exist with remote cache feature

### LINKED ISSUE #30855: Make persistent user sessions and external Infinispan co-exist
### Description

The durable user sessions developed in #28265 should work together with #28745.

Acceptance criteria (draft): 
* If persistent user sessions and external Infinispan are enabled, the user sessions should be handled by the the persistent user sessions code.
* If external Infinispan is enabled, the persistent user sessions shouldn't use the external Infinispan for caching sessions.

### Motivation

Be able to provide an A/A Multi-site persistent session w/ external Infinispan.

### Acceptance criteria

* [x] Implement a switch to disable caching in single-site setup to allow for performance comparison
* [x] Do not use any caches for user sessions in a multi-site setup as they can't be kept up-to-date in a A/A setup
* [ ] Performance test in KCB should show no regression
* [ ] Functional test in KCB green: A/P remote cache -> A/A remote cache + persistent sessions (waiting for Pedro's PR)

### Details

_No response_

### PR DESCRIPTION
**Pull request by ahus1**
Closes #30855

Results with 50 logins+logouts per second, single-site ROSA, single site Aurora -> almost no change in the response times

Single user (so all other data is in memory), 3 pods, 4 CPU requested, 5 min run, ~1,5 vCPU per pod, Aurora 16.1

```
./benchmark.sh eu-west-1 --scenario=keycloak.scenario.authentication.AuthorizationCode --server-url=${KEYCLOAK_URL} --realm-name=test-realm --ramp-up=20 --measurement=300 --clients-per-realm=1 --users-per-realm=1 --users-per-sec=50 --logout-percentage=100 --log-http-on-failure
```

Scenario: volatile / persistent / persistent nocache

CPU usage database: 13 / 33 / 36

Commit count: 38 / 250 / 360

image

image

image

### CODE REVIEW

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/PersistentUserSessionProvider.java [hunk 1]
```diff
@@ -425,6 +425,10 @@ public UserSessionModel getUserSessionWithPredicate(RealmModel realm, String id,
 
         // Try lookup userSession from remoteCache
```
Comment (pruivo) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/PersistentUserSessionProvider.java:426]: I don't understand this method's logic. Line 421, `predicate.test(userSession)`, if the predicate returns `false`, why isn't the method returning `null`? Why does it continue to try to fetch the session from the cache? Are we expecting the predicate to return something different with the cached data?

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsChangelogBasedTransaction.java [hunk 2]
```diff
@@ -66,46 +63,49 @@ public PersistentSessionsChangelogBasedTransaction(KeycloakSession session,
                                                        SerializeExecutionsByKey<K> serializerOffline) {
         kcSession = session;
 
-        if (!Profile.isFeatureEnabled(Profile.Feature.PERSISTENT_USER_SESSIONS)) {
```
Comment (mhajas) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsChangelogBasedTransaction.java:69]: Why is this removed?

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProviderFactory.java
Comment (mhajas) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProviderFactory.java:1]: I am wondering whether it would make sense to create a new Factory for the persistent sessions without Infinispan. Could it simplify things? We can be then sure no embedded Infinispan is used when persistent session and multisite is used. 

The idea would be to have a different Factory enabled based on the flag, essentially this Factory would be used the same way as it was prior to this PR for single site deployments while the second without any Infinispan code could be used for the multisite. 

WDYT?

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/ClientSessionPersistentChangelogBasedTransaction.java [hunk 3]
```diff
@@ -55,15 +57,17 @@ public ClientSessionPersistentChangelogBasedTransaction(KeycloakSession session,
...
+        super(session, CLIENT_SESSION_CACHE_NAME, cache, offlineCache, remoteCacheInvoker, lifespanMsLoader, maxIdleTimeMsLoader, offlineLifespanMsLoader, offlineMaxIdleTimeMsLoader, batchingQueue, serializerOnline, serializerOffline);
         this.userSessionTx = userSessionTx;
     }
 
     public SessionEntityWrapper<AuthenticatedClientSessionEntity> get(RealmModel realm, ClientModel client, UserSessionModel userSession, UUID key, boolean offline) {
         SessionUpdatesList<AuthenticatedClientSessionEntity> myUpdates = getUpdates(offline).get(key);
         if (myUpdates == null) {
             SessionEntityWrapper<AuthenticatedClientSessionEntity> wrappedEntity = null;
-            wrappedEntity = getCache(offline).get(key);
+            if (getCache(offline) != null) {
```
Comment (mhajas) [inline model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/ClientSessionPersistentChangelogBasedTransaction.java:68]: nitpick: here we can do one less call to `getCache` method if we store it in some variable

**Review by mhajas — COMMENTED** [review]
Thank you for the PR @ahus1, I added a few comments. Otherwise it LGTM

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsChangelogBasedTransaction.java (hunk 2 above)
Comment (ahus1) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/PersistentSessionsChangelogBasedTransaction.java:69]: While this is a safeguard, IMHO it is never happening as we're creating this transaction only when persistent user sessions are enabled.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProviderFactory.java
Comment (ahus1) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/InfinispanUserSessionProviderFactory.java:1]: IMHO this would be a good follow-up. I would then check what code is common code and move that to a base class, and which is specific to each implementation.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/PersistentUserSessionProvider.java (hunk 1 above)
Comment (ahus1) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/PersistentUserSessionProvider.java:426]: True, this is strange and it is worth another analysis. Please create a follow-up issue as it is IMHO unrelated to this issue.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/ClientSessionPersistentChangelogBasedTransaction.java (hunk 3 above)
Comment (ahus1) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/changes/ClientSessionPersistentChangelogBasedTransaction.java]: Addressed in the second commit for this issue.

File: model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/PersistentUserSessionProvider.java (hunk 1 above)
Comment (ahus1) [reply model/infinispan/src/main/java/org/keycloak/models/sessions/infinispan/PersistentUserSessionProvider.java:426]: Issue created: https://github.com/keycloak/keycloak/issues/31115

**Review by pruivo — APPROVED** [review]

**Review by mhajas — APPROVED** [review]
Thank you @ahus1!
